### PR TITLE
Do not use ClusterRole and ClusterRoleBinding when .Values.mgmt.namespaces list is empty

### PR DESCRIPTION
**Pull request by AndreaVida**
Support watching all namespaces (*) or only one with limited privileges (no ClusterRole and ClusterRoleBinding).

Default values are backward compatible.

Closes https://github.com/open-policy-agent/kube-mgmt/issues/212.

### CODE REVIEW

**Review by eshepelyuk — COMMENTED** [review]
General comments

1. plz rebase and squash into single commit.
1. PR should be covered with unit test.
1. regarding e2e tests i think there is no meed to create a new one, existing should work.

Implementation.

There is no reason in introducing a new parameter, this can be covered by existing parameter `namespace`.  By default, current namespace is watched, this behaviour should be kept. The Role template should iterate over configured namespaces and create a role for every namespace, unless namespaces is [*], in this case a single cluster role should be created.

**Comment by eshepelyuk** [comment]
Also remember, there is a separate setting for replication namespaces, so those namespaces should also be taken into account when creating `Role`.

**Comment by AndreaVida** [comment]
Hi @eshepelyuk,

about unit tests; my changes are only on the Helm Chart and I honestly have no idea on how unit tests can be done on it.

About implementation, I will try to not add a new value if possible and default behavior of watching current namespace is kept already. 
I am not 100% sure about the roles iteration stuff though, as the Go K8s library only supports one or all namespaces use cases and I am afraid the all case may require privileges to access all. Do you know more on this?

**Comment by eshepelyuk** [comment]
> Hi @eshepelyuk,
> 
> about unit tests; my changes are only on the Helm Chart and I honestly have no idea on how unit tests can be done on it.
> 
You should check `test/unit` folder and justfile recipe `test-helm`

> About implementation, I will try to not add a new value if possible and default behavior of watching current namespace is kept already. 
> I am not 100% sure about the roles iteration stuff though, as the Go K8s library only supports one or all namespaces use cases and I am afraid the all case may require privileges to access all. Do you know more on this?  

I was referring to **helm template** not golang code.

**Comment by AndreaVida** [comment]
> I was referring to **helm template** not golang code.

Indeed, but if the library requires cluster-wide access when watching on `Namespaces.All` it will not work without a `ClusterRole`. That's the thing I am not sure about. Are you?

**Comment by eshepelyuk** [comment]
> > I was referring to **helm template** not golang code.
> 
> Indeed, but if the library requires cluster-wide access when watching on `Namespaces.All` it will not work without a `ClusterRole`. That's the thing I am not sure about. Are you?

But if watched namespaces has value of * -the ClusterRole should be gemerated by helm, as now.

**Comment by AndreaVida** [comment]
> > > I was referring to **helm template** not golang code.
> > 
> > 
> > Indeed, but if the library requires cluster-wide access when watching on `Namespaces.All` it will not work without a `ClusterRole`. That's the thing I am not sure about. Are you?
> 
> But if watched namespaces has value of * -the ClusterRole should be gemerated by helm, as now.

Exactly, I was referring to the case when namespaces list length is > 1 but I just double checked golang code and seems like in that case one Informer instance is spawned for each namespace to watch; if that's the case then your proposal should work.
Do you confirm this is how it works?

**Comment by eshepelyuk** [comment]
yes I do.

**Comment by AndreaVida** [comment]
Thanks for the clarification! I will then work on discussed changes from now on.

File: charts/opa-kube-mgmt/templates/deployment.yaml [hunk 1]
```diff
@@ -181,7 +181,11 @@ spec:
             {{- end }}
             - --opa-url={{ .Values.useHttps | ternary "https" "http" }}://127.0.0.1:{{ .Values.port }}/v1
             - --opa-allow-insecure
+            {{- if has "*" .Values.mgmt.namespaces }}
```
Comment (eshepelyuk) [inline charts/opa-kube-mgmt/templates/deployment.yaml:184]: There is no need to introduce new if block, current code covers all cases already.

File: charts/opa-kube-mgmt/templates/rbac-mgmt.yaml [hunk 2]
```diff
@@ -1,4 +1,6 @@
 {{- if .Values.rbac.create }}
+
+{{- if has "*" .Values.mgmt.namespaces }}
```
Comment (eshepelyuk) [inline charts/opa-kube-mgmt/templates/rbac-mgmt.yaml:3]: Should be equality check .

**Comment by eshepelyuk** [comment]
Unfortunately there is a deeper issue.
ClusterRole is mandatory for k8s resources replication, so it must be created as soon as resource replication is required.

So much deeper changes required. 
The should be separate roles for policy \ data loading and for resource replication.
My suggestion is to create two separate templates for them.

**Pull request closed without merging**